src/scheduler/gc_work.rs

```diff
@@ -446,39 +450,36 @@ impl<E: ProcessEdgesWork> GCWork<E::VM> for ScanObjects<E> {
     }
 }
 
-#[derive(Default)]
 pub struct ProcessModBuf<E: ProcessEdgesWork> {
-    modified_nodes: Vec<ObjectReference>,
-    modified_edges: Vec<Address>,
+    modbuf: Vec<ObjectReference>,
     phantom: PhantomData<E>,
+    meta: SideMetadataSpec,
 }
 
 impl<E: ProcessEdgesWork> ProcessModBuf<E> {
-    pub fn new(modified_nodes: Vec<ObjectReference>, modified_edges: Vec<Address>) -> Self {
+    pub fn new(modbuf: Vec<ObjectReference>, meta: SideMetadataSpec) -> Self {
         Self {
-            modified_nodes,
-            modified_edges,
+            modbuf,
+            meta,
             phantom: PhantomData,
         }
     }
 }
 
 impl<E: ProcessEdgesWork> GCWork<E::VM> for ProcessModBuf<E> {
-    #[inline]
+    #[inline(always)]
     fn do_work(&mut self, worker: &mut GCWorker<E::VM>, mmtk: &'static MMTK<E::VM>) {
+        if !self.modbuf.is_empty() {
+            for obj in &self.modbuf {
+                compare_exchange_atomic(self.meta, obj.to_address(), 0b0, 0b1);
+            }
+        }
         if mmtk.plan.in_nursery() {
-            let mut modified_nodes = vec![];
-            ::std::mem::swap(&mut modified_nodes, &mut self.modified_nodes);
-            worker.scheduler().work_buckets[WorkBucketStage::Closure]
-                .add(ScanObjects::<E>::new(modified_nodes, false));
-
-            let mut modified_edges = vec![];
-            ::std::mem::swap(&mut modified_edges, &mut self.modified_edges);
-            worker.scheduler().work_buckets[WorkBucketStage::Closure].add(E::new(
-                modified_edges,
-                true,
-                mmtk,
-            ));
+            if !self.modbuf.is_empty() {
+                let mut modbuf = vec![];
+                ::std::mem::swap(&mut modbuf, &mut self.modbuf);
+                GCWork::do_work(&mut ScanObjects::<E>::new(modbuf, false), worker, mmtk)
```

Comment:
This will do the work with the current thread and the current work packet. What is the reason for this change? Does  this affect the correctness of work packet statistics? Basically the `ScanObject` work is counted in a `ProcessModBuf` work.

Reply:
This work-packet only produces one (instead of `N`) ScanObjects packets. So I think it makes no sense to push it to the local work-packet cache and pull it pack before executing it.

Yes it will affect the statistics. Thanks for pointing this out. I'll fix this.

Reply:
I double checked the code. By calling `GCWork::do_work(&mut ScanObjects...)` directly, the `ScanObjects` packet is not accounted when collecting statistics.